### ISSUE TITLE
Add IPv6 TCP Server Endpoint

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5694]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)| @ashfall reported|
|-|-|
|Trac ID|trac#5694|
|Type|enhancement|
|Created|2012-05-30 15:18:08Z|
|Branch|https://github.com/twisted/twisted/tree/ipv6-tcp-server-endpoint-5694|

Narrowing down [#4470](https://github.com/twisted/twisted/issues/4470) for servers.

<details><summary>Searchable metadata</summary>

```
trac-id__5694 5694
type__enhancement enhancement
reporter__ashfall ashfall
priority__normal normal
milestone__ 
branch__branches_ipv6_tcp_server_endpoint_5694 branches/ipv6-tcp-server-endpoint-5694
branch_author__ashfall ashfall
status__closed closed
resolution__fixed fixed
component__core core
keywords__endpoint_ipv6 endpoint ipv6
time__1338391088000000 1338391088000000
changetime__1339165316000000 1339165316000000
version__None None
owner__ashfall ashfall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_1"></a>|@ashfall commented|
|-|-|

(In [34469]) Branching to 'ipv6-tcp-server-endpoint-5694'

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_2"></a>|@ashfall removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ipv6-tcp-server-endpoint-5694)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks!

  1. The news fragment might as well explicitly name the new API.  The goal of these news fragments is to make it easy to see what's changed in a new release of Twisted.
  1. It's unfortunate to have to document `_port`/`port`, `_reactor`/`reactor`, etc on `TCP6ServerEndpoint` twice.  I see that the other endpoints do this as well.  I think it's a mistake.  Documenting only the `__init__` parameters, so long as they very closely mirror the attributes, should be sufficient.
  1. Also, the `_listenArgs` attribute is defined but never used.  This is also true for the same attribute on `TCP4ServerEndpoint`.  And `UNIXServerEndpoint` documents it but doesn't define it.
  1. The only difference between `TCP6ServerEndpoint` and `TCP4ServerEndpoint` seems to be the default value for `interface`.  Perhaps there's a way to implement this without duplicating so much code?
  1. I see a bunch of code about clients in the unit tests, but there's no client functionality being added in this branch.  Perhaps this code belongs in the client branch, instead?
  1. The typical way of spelling "0:0:0:0:0:0:0:1" is "::1"
  1. We also need a string endpoint description plugin for this new server endpoint.  Can you file a new ticket for that?  Generally, there should be one of these for every kind of endpoint.  Without them, the endpoints aren't accessible from the command line.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_4"></a>|@ashfall removed owner|
|-|-|

Changed code as per exarkun's review.

Filed [#5699](https://github.com/twisted/twisted/issues/5699) for the string endpoint description plugin.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks!  A few more comments:

  1. `TCP4ServerEndpoint` and `TCP6ServerEndpoint` can both be simplified.  The `listen` method they both implement is identical and provides exactly the same behavior as the base class's `listen` method.  If you delete the method from both subclasses, they'll keep working the same.
  1. `TCP6ServerEndpoint` also doesn't need to declare it implements `IStreamServerEndpoint`.  It inherits that declaration from the base class.
  1. The introduction of `TCPServerEndpoint` raises some questions.  As an approach to code sharing, it's fine and does the job well.  As a public interface, though, it's redundant with `TCP4ServerEndpoint` and `TCP6ServerEndpoint`.  I think having both `TCP4ServerEndpoint` and `TCP6ServerEndpoint` _also_ presents some redundancy, but that's the plan we've chosen so I guess I shouldn't be worrying about it now.
  1. You can use _@type_ in an `__init__` docstring.  Please do that instead of starting _@param_ fields with _(some type)_.
  1. The change to `test_endpointListenSuccess` partly breaks the intent of the test, which is to exercise the defaults for the endpoint.  Previously `backlog` and `interface` were not passed, but the test verified that the proper defaults were passed to `listenTCP`.  Now the test exercises basically the same code path as `test_endpointListenNonDefaultArgs`. Fortunately the change doesn't seem necessary?  The tests pass on my laptop if I revert it.  Did I overlook something?
  1. The docstring for `createServerEndpoint` says "an L{TCP6ServerEndpoint}".  Should be "a L{TCP6ServerEndpoint}", since the "L" is silent. :)
  1. `twisted/test/proto_helpers.py` still has this pyflakes warning: `twisted/test/proto_helpers.py:415: undefined name 'UnsupportedAddressFamily'`.  Can you also fix that one?
  1. The news fragment is a little awkward.  Move all of the name up front and it will flow a little better: _twisted.internet.endpoints now provides TCP6ServerEndpoint, an IPv6 TCP server endpoint._

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_6"></a>|@ashfall removed owner|
|-|-|

Addressed the issues raised in exarkun's review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ipv6-tcp-server-endpoint-5694)

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_8"></a>|@ashfall commented|
|-|-|

Also fixes [#5692](https://github.com/twisted/twisted/issues/5692).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @ashfall|
|-|-|

Thanks.  Regarding point 3 on my last review, I think `TCPServerEndpoint` should be made private instead (which I wasn't clear on in my last review, sorry about that).  Please rename it `_TCPServerEndpoint` and remove it from `__all__`.  After that I think the branch will be fine to merge.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_10"></a>|@ashfall set status to `closed`|
|-|-|

(In [34526]) Merge ipv6-tcp-server-endpoint: Added IPv6 TCP Server Endpoint.

Author: ashfall
Reviewer: exarkun
Fixes: [#5692](https://github.com/twisted/twisted/issues/5692), [#5694](https://github.com/twisted/twisted/issues/5694)

This adds an IPv6 stream endpoint through twisted.internet.endpoints.TCP6ServerEndpoint.
